compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3109,6 +3124,31 @@ BType getTypeIntersection(BType lhsType, BType rhsType) {
                 }
             } else if (type.tag == TypeTags.NULL_SET) {
                 return type;
+            } else if (type.tag == TypeTags.ERROR && lhsType.tag == TypeTags.ERROR) {
+                BType intersectionType = getIntersectionForErrorTypes(lhsType, type, pkgEnv);
+                if (intersectionType != symTable.semanticError) {
+                    return intersectionType;
+                }
+            } else if (type.tag == TypeTags.RECORD && lhsType.tag == TypeTags.RECORD) {
+                BType intersectionType = createRecordIntersection(lhsType, type, pkgEnv);
+                if (intersectionType != symTable.semanticError) {
+                    return intersectionType;
+                }
+            } else if (type.tag == TypeTags.MAP && lhsType.tag == TypeTags.RECORD) {
+                BType intersectionType = createRecordAndMapIntersection(lhsType, type, pkgEnv);
+                if (intersectionType != symTable.semanticError) {
+                    return intersectionType;
+                }
+            } else if (type.tag == TypeTags.RECORD && lhsType.tag == TypeTags.MAP) {
+                BType intersectionType = createRecordAndMapIntersection(type, lhsType, pkgEnv);
+                if (intersectionType != symTable.semanticError) {
+                    return intersectionType;
+                }
+            } else if (type.tag == TypeTags.MAP && lhsType.tag == TypeTags.MAP) {
+                BType intersectionType = createRecordAndMapIntersection(type, lhsType, pkgEnv);
```

Comment:
Is this method name correct? IMO, under this condition we cannot call `createRecordAndMapIntersection`, we need to call another method to get the intersection between two maps. WDYT?

Reply:
The idea was to create a same method to get the interesctions of records, as well as maps. I too feel this is kinda confusing. I'll check how I can improve this.